File: PhaseJump/PhaseJump/Classes/Interface/ButtonControl.cpp
#include "ButtonControl.h"
#include "SomeStateHandler.h"

using namespace std;
using namespace PJ;

void ButtonControl::SetIsHovering(bool value)
{
    if (isTracking) { return; }
    isHovering = value;
    core->SetState(isHovering ? StateType::Hover : StateType::Normal);
}

ButtonControl::ButtonControl()
{
}

void ButtonControl::Awake() {
    Base::Awake();

    auto sharedThis = SCAST<ButtonControl>(shared_from_this());
    core->Configure(SCAST<SomeGoStateListener<ButtonControlStateType>>(sharedThis));
}

void ButtonControl::OnPointerDownEvent(PointerDownUIEvent<LocalPosition> _event)
{
    switch (trackingType)
    {
        case TrackingType::Immediate:
            OnPress();
            break;
        case TrackingType::Track:
            isTracking = true;
            core->SetState(StateType::Press);
            break;
    }
}

void ButtonControl::OnPointerEnterEvent(PointerEnterUIEvent _event)
{
    if (!isTracking) { return; }
    core->SetState(StateType::Press);
}

void ButtonControl::OnPointerExitEvent(PointerExitUIEvent _event)
{
    if (!isTracking) { return; }
    core->SetState(StateType::Normal);
}

void ButtonControl::OnPointerUpEvent(PointerUpUIEvent _event)
{
    if (!isTracking) { return; }
    isTracking = false;

    bool wasPressed = false;
    switch (core->State())
    {
        case StateType::Press:
            core->SetState(isHovering ? StateType::Hover : StateType::Normal);
            wasPressed = true;
            break;
        default:
            core->SetState(StateType::Normal);
            break;
    }

    if (wasPressed)
    {
        OnPress();
    }
}

void ButtonControl::OnPress()
{
    // Debug.Log("Button Pressed");
    auto listener = Listener();
    if (nullptr == listener) { return; }
    listener->OnEvent(MAKE<PressEvent>(id));
}

void ButtonControl::OnStateChange(GoStateMachine<StateType>& inStateMachine)
{
    auto owner = this->owner.lock();

    // Forward button state change to state handler components
    auto stateHandlers = owner->TypeComponentsInChildren<SomeStateHandler<StateType>>();
    for (auto stateHandler : stateHandlers)
    {
        stateHandler->OnStateChange(inStateMachine.State());
    }
}

void ButtonControl::OnStateFinish(GoStateMachine<StateType>& inStateMachine)
{
}

File: Experimental/PhaseJumpUnity/Assets/phasejumppro/UI/Controls/ButtonControl.cs
using UnityEngine;
using UnityEngine.EventSystems;

/*
RATING: 5 stars
Tested and works
CODE REVIEW: 6/10/23
PORTED TO: C++
*/
namespace PJ
{
    // FUTURE: ButtonView wraps this
    public class ButtonControl : UIControl2D, SomeGoStateListener<ButtonControl.StateType>
    {
        public class PressEvent : Event
        {
            public PressEvent(string id) : base(id)
            {
            }
        }

        /// <summary>
        /// Button state
        /// IMPORTANT: Keep state IDs consistent for serialization
        /// </summary>
        public enum StateType
        {
            // Default
            Normal = 0,

            // Pointer (mouse) is hovering over
            Hover = 1,

            // Is pressed while tracking pointer input
            Press = 2,

            // Disabled
            Disabled = 3
        }

        public enum TrackingType
        {
            // Track the button
            Track,

            // Fire the button press on tap
            Immediate
        }

        public override bool IsHovering
        {
            set
            {
                isHovering = value;
                if (isTracking) { return; }

                core.State = isHovering ? StateType.Hover : StateType.Normal;
            }
        }

        [Header("Button Properties")]

        public TrackingType trackingType = TrackingType.Track;

        protected GoCore<StateType> core;
        protected bool isTracking = false;

        public ButtonControl()
        {
            this.core = new(this);
        }

        public override void OnPointerDownEvent(PointerDownUIEvent<LocalPosition> _event)
        {
            switch (trackingType)
            {
                case TrackingType.Immediate:
                    OnPress();
                    break;
                case TrackingType.Track:
                    isTracking = true;
                    core.State = StateType.Press;
                    break;
            }
        }

        public override void OnPointerEnterEvent(PointerEnterUIEvent _event)
        {
            if (!isTracking) { return; }
            core.State = StateType.Press;
        }

        public override void OnPointerExitEvent(PointerExitUIEvent _event)
        {
            if (!isTracking) { return; }
            core.State = StateType.Normal;
        }

        public override void OnPointerUpEvent(PointerUpUIEvent _event)
        {
            if (!isTracking) { return; }
            isTracking = false;

            bool wasPressed = false;
            switch (core.State)
            {
                case StateType.Press:
                    core.State = isHovering ? StateType.Hover : StateType.Normal;
                    wasPressed = true;
                    break;
                default:
                    core.State = StateType.Normal;
                    break;
            }

            if (wasPressed)
            {
                OnPress();
            }
        }

        /// <summary>
        /// Called when the button is pressed
        /// </summary>
        protected virtual void OnPress()
        {
            // Debug.Log("Button Pressed");
            var listener = Listener;
            if (null == listener) { return; }
            listener.OnEvent(new PressEvent(id));
        }

        public void OnStateChange(GoStateMachine<StateType> inStateMachine)
        {
            // Forward button state change to state handler components
            var stateHandlers = GetComponentsInChildren<SomeStateHandler<ButtonControl.StateType>>();
            foreach (var stateHandler in stateHandlers)
            {
                stateHandler.OnStateChange(inStateMachine.State);
            }
        }

        public void OnStateFinish(GoStateMachine<StateType> inStateMachine)
        {
        }
    }
}
